src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditWaiter.cs

```diff
@@ -10,21 +10,29 @@
 namespace System.Net.Http
 {
     /// <summary>Represents a waiter for credit.</summary>
-    internal class CreditWaiter : IValueTaskSource<int>
+    internal sealed class CreditWaiter : IValueTaskSource<int>
     {
+        private CancellationTokenRegistration _registration;
+        private ManualResetValueTaskSourceCore<int> _source;
+        private int _completionReserved;
+        private volatile bool _registrationSet;
```

Comment:
> but could you replace _completionReserved and _registrationSet with a field that stores _source.Version?

No, it would not have the right atomicity required. These fields are also unrelated; one is about ensuring that only one thread gets permission to transition the instance to completed, the other is about ensuringb that cancellation only sees a fully-formed registration.